File: byteme-main/go/composite.go
package main

import("os"
       "fmt"
	"strconv"
	"math")

//Precondition: user has passed in upper and lower bound through command line
//Postcondition: Vector of size size upper_bound is filled with 0's and 1's, wi
//		with a 1 indicvation value of the index is a composite integer

func gen_composites(n int64) int64{
		
	
	//Precondition: an integer between upper and lower bounds defined by user is passed in
	//Postcondition: a 0 is returned if n is prime and a 1 is returned if n is composite

	//2 and 3 are prime, return 0
	if n == 2 || n == 3{
		return 0	
	}

	//if n is divisble by 2 or 3 it is composite, return 1
	if n <= 1 ||  n % 2 == 0 || n % 3 == 0{
		return 1
	} 
	
	i := int64(5)
			
	//if n has factors other than itself, it is composite, return 1
	for i*i <= n{
		if n % i == 0 || n % (i + 2) == 0{
			return 1
		}
		
		i += 6
	}
	
	//if all conditions fail, n is prime, return 0
	return 0	
}

//Precondition: Sum of all prime factors of all valid composites is passed in
//		to function
//Recursive calls take value from previous call
//Postcondition: Sum is collapsed into a single digit

func collapse(num int64) int64{
	
	//Set temp value for collapse to num parameter
	temp_collapse := num
	
	//Base Case: If num is a single digit, break recursion and return the current num
	if (num % 10) != num{
		
		//While num is not a single digit, add all digits of num and recursively
		//call collapse on that sum
		temp_collapse = 0
		
		for num % 10 != num{
			
			//num % 10 gives last digit of num, num/10 returns num witout
			//last digit
			temp_collapse += (num % 10)
			num = num/10
		}
	
		//add final digit to the collapse sum
		temp_collapse += num
		
		//recursive call of collapse
		temp_collapse = collapse(temp_collapse)

	}
			
	return temp_collapse

}


//Precondition: Collapse of sum of valid prime factors is passed into function
//Postcondition: Roman numeral corresponding to collapse is printed
func roman_numeral(n int64) string{
	//If collapse does not match a roman numeral from 1-9, N is printed
	switch n{
	case 1:
		return "I"
	case 2:
		return "II"
	case 3:
		return "III"
	case 4:
		return "IV"
	case 5:
		return "V"
	case 6:
		return "VI"
	case 7:
		return "VII"
	case 8:
		return "VIII"
	case 9:
		return "IX"
	default:
		return "N"
	}
}


//Main driver code
func main(){
	//Collecting lower and upper bounds from command line
	//assumes lower bound is entered first and upper bound second
	lower_bound, err:= strconv.ParseInt(os.Args[1], 10, 64)
	upper_bound, err := strconv.ParseInt(os.Args[2], 10, 64)
	_ = err

	var collapse_sum int64	
	//variable for storing sum of primes
	var sum int64
	
	//counter for number of composites with 2 unique prime factors
	var valid_count int64
	
	//array for storing prime factors of composites with unique in range of upper and lower bounds
	var valid_primes[] int64

	//array for storing all prime factors in upper and lower bounds
	var prime_factors[] int64


	

	var j int64
	var i int64	
	//Determining which composites in range are valid using prime factorization
	//checking for composites with only two unique prime factors
	//for loops upper bound is exclusive, so loops to upper bound + 1
	for i = lower_bound; i < upper_bound + 1; i++{
			
		//If value in composite vector is one, perform factorization on that index
		if gen_composites(i) == 1{
			
			target_composite := i
					
			prime_factors = nil	
			//Perform prime factorization on target composite
			//loop from 2 to squareroot of composite
			root := int64(math.Ceil(math.Sqrt(float64(i))))
			for j = 2; j < root; j++ {
				
				if  target_composite != 1{
					
					for target_composite % j == 0{
						
						if len(prime_factors) < 4{
							prime_factors = append(prime_factors, j)
							target_composite = target_composite / j
												
						}else{
							break
						}		
					}	
				}
			
			}
					
				
			//if last factor is prime, append to prime_factors
			if gen_composites(target_composite) == 0{
				prime_factors = append(prime_factors, target_composite)
			} 		
			
			//If array of prime facotrs contains 2 values that are unique, add those factors to array
			//of all valid prime factors and increment count of valid composites		
			if len(prime_factors) == 2 && prime_factors[0] != prime_factors[1]{
				valid_primes = append(valid_primes, prime_factors[0])
				valid_primes = append(valid_primes, prime_factors[1])
				valid_count += 1
				
			}
		}
	}
	
	for i := 0; i < len(valid_primes); i++{
		sum += valid_primes[i]
	}	

	if len(valid_primes) > 0{
		collapse_sum = collapse(sum)
	}
	
		
	fmt.Println("Count:", valid_count)
	fmt.Println("Sum:", sum)
	fmt.Println("Collapse:", collapse_sum)
	fmt.Println("Roman:", roman_numeral(collapse_sum))				
	fmt.Println("Hail Caesar!")	
	


}	

File: byteme-main/rust/composite.rs
#![recursion_limit="20000"]
use std::env;

//Precondition: Integer is passed in through main loop of program
//Postcondition: Returns boolean indicating if input value is composite
fn is_composite(num:i128) -> bool
{
	//Uses primality test to check if input value is composite

	//Variable for while loop to check primality of input value
	let mut loop_counter: i128 = 5;

	//If input value is 2 or 3, return false
	if num == 2 || num == 3 {
		return false;
	}
	
	//If input value is even or divisible by 3, return true
	if num <= 1 || num % 2 == 0 || num % 3 == 0 {
		return true;
	}

	//Test if input value has factors other than itself and 1. If so, return true
	while loop_counter.pow(2) <= num {

		if num % loop_counter == 0 || num % (loop_counter + 2) == 0 {
			return true;
		}

		loop_counter += 6;
	}

	//If all previous tests fail, return false
	return false;
}

//Precondition: Sum of all prime factors of valid composites is passed in to function
//Recursive calls take value from previous call
//Postcondition: Sum is collapsed into a single digit
fn collapse(mut num: i128)->i128 {

	//Set temp value for collapse to num parameter
	let mut temp_collapse:i128 = num;

	//Base case: If num is a single digit, stop recursion and return current num
	if num % 10 != num {

		//While num is not a single digit, add all digits of num and recursively
		//call collapse on that sum
		temp_collapse = 0;
		while num % 10 != num {
			//num % 10 gives last digit of num, num / 10 returns num w/o last digit
			temp_collapse += num % 10;
			num = num / 10;
		}
		//Add final digit (first digit of the initial num) to the collapse sum
		temp_collapse += num;	

		//Recursive call of collapse
		temp_collapse = collapse(temp_collapse);
	}
	
	//Returning collapse sum
	return temp_collapse;
}

//Precondition: Collapse of sum of valid prime factors is passed into function
//Postcondition: Roman numeral corresponding to collapse is printed
fn roman_numeral(num: i128) {
	//Roman numeral equivalent to collapse sum is printed
	//If collapse does not match a roman numeral from 1 - 9, prints N
	match num {
	1=> println!("Roman: I"),
	2=> println!("Roman: II"),
	3=> println!("Roman: III"),
	4=> println!("Roman: IV"),
	5=> println!("Roman: V"),
	6=> println!("Roman: VI"),
	7=> println!("Roman: VII"),
	8=> println!("Roman: VIII"),
	9=> println!("Roman: IX"),
	_=> println!("Roman: N"),
	}
}

//Main driver code
fn main()
{
	//Collecting command line arguments and setting lower and upper bounds
	//Assumes lower bound is entered first and upper bound is entered second
	let args: Vec<String> = env::args().collect();
	let lower_bound: i128 = args[1].parse().unwrap();
	let upper_bound: i128 = args[2].parse().unwrap();

	//Counter for the number of composites in range w/ two unique prime factors
	let mut valid_count: i32 = 0;	

	//Vector for storing prime factors of every valid composite w/ two unique prime factors
	let mut valid_primes = vec![0;0];

	//Variable for storing composite to perform factorization on
	let mut target_composite: i128 ;

	//Vector for storing all the prime factors of a given composite in given range
	let mut prime_factors = vec![0;0];

	//Determining which composites in range are valid using prime factorization and
	//checking for those with only two unique prime factors
	//For loop's upper bound is exclusive, so running from lower_bound to upper_bound + 1
	for i in lower_bound..upper_bound+1 {
	
		//If value in range is composite, sets it as target to perform prime factorization on
		if is_composite(i) == true {
			target_composite = i;	

			//Emptying vector of prime factors for storing target composite's prime factors
			while !prime_factors.is_empty() {
				prime_factors.pop();
			}

			//Perform prime factorization on target composite
			for j in 2..upper_bound {

				//If vector of prime factors contains 3 factors, conditions have not been met, so
				//end factorization
				while target_composite % j == 0 && prime_factors.len() < 4 {
					target_composite = target_composite / j;

					//Add factor to vector of prime factors
					prime_factors.push(j);
				}
			}

			//If vector of prime factors contains 2 values that are unique, add those factors to vector
			//of all valid prime factors and increment count of valid composites
			if prime_factors.len() == 2 && prime_factors[0] != prime_factors[1] {
				valid_primes.push(prime_factors[0]);
				valid_primes.push(prime_factors[1]);
				valid_count += 1;
			}
		}
	} 

	//Value for storing sum of all valid prime factors
	let mut sum: i128 = 0;

	//Sum all valid prime factors in vector
	for i in 0..valid_primes.len() {
		sum += valid_primes[(i as usize)];
	}
	
	//Print count and sum of valid prime factors
	println!("Count: {:?}", valid_count);
	println!("Sum: {:?}", sum);

	//Variable for storing collapse of sum of prime factors
	let mut collapse_sum: i128 = 0;
	
	//If there are any valid primes, perform collapse on sum
	if valid_primes.len() > 0 {
		collapse_sum = collapse(sum);
	}

	//Printing collapse of sum of prime factors
	println!("Collapse: {:?}", collapse_sum);

	//Printing roman numeral corresponding to collapse
	roman_numeral(collapse_sum);

	//Printing "Hail Caesar!"
	println!("\t-- Hail Caesar!");	
}

